src/callbacks_step/analysis_dg2d_parallel.jl

```diff
@@ -12,31 +12,92 @@ function calc_error_norms(func, u, t, analyzer,
-  total_volume = mesh.tree.length_level_0^ndims(mesh)
-  global_l2_error = Vector(l2_error.^2 .* total_volume)
-  global_linf_error = Vector(linf_error)
-  MPI.Reduce!(global_l2_error, +, mpi_root(), mpi_comm())
-  MPI.Reduce!(global_linf_error, max, mpi_root(), mpi_comm())
+  l2_errors, linf_errors = calc_error_norms_per_element(func, u, t, analyzer,
+                                                        mesh, equations, initial_condition,
+                                                        dg, cache, cache_analysis)
+
+  nvars = length(l2_errors[1])
+  T = eltype(l2_errors)
+
+  # Convert errors from Vector of SVectors to Matrix for MPI communication
+  FT = eltype(l2_errors[1])
+  l2_errors = reshape(reinterpret(FT, l2_errors), nvars, :) |> Matrix
+  linf_errors = reshape(reinterpret(FT, linf_errors), nvars, :) |> Matrix
+
+  # Collect local error norms of all elements on root process
   if mpi_isroot()
-    l2_error   = convert(typeof(l2_error),   global_l2_error)
-    linf_error = convert(typeof(linf_error), global_linf_error)
+    global_l2_errors = zeros(eltype(l2_errors), nvars, cache.mpi_cache.n_elements_global)
+    global_linf_errors = similar(global_l2_errors)
+
+    n_elements_by_rank = parent(cache.mpi_cache.n_elements_by_rank) # convert OffsetArray to Array
+    l2_buf = MPI.VBuffer(global_l2_errors, nvars*n_elements_by_rank)
+    linf_buf = MPI.VBuffer(global_linf_errors, nvars*n_elements_by_rank)
+    MPI.Gatherv!(l2_errors, l2_buf, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, linf_buf, mpi_root(), mpi_comm())
   else
-    l2_error   = convert(typeof(l2_error),   NaN * global_l2_error)
-    linf_error = convert(typeof(linf_error), NaN * global_linf_error)
+    MPI.Gatherv!(l2_errors, nothing, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, nothing, mpi_root(), mpi_comm())
   end
 
-  l2_error = @. sqrt(l2_error / total_volume)
+  # Aggregate element error norms on root process
+  if mpi_isroot()
+    # Convert from Matrix to Vector of SVectors
+    global_l2_errors = reinterpret(T, vec(global_l2_errors))
+    global_linf_errors = reinterpret(T, vec(global_linf_errors))
+
+    # Aggregate element errors
+    # sum(global_l2_errors) does not produce the same result as in the serial case, while
+    # sum(global_l2_errors[:]) produces the same results but leads to unnecessary allocations.
+    # Therefore we use a hand-written loop here
+    l2_error = zero(T)
+    for error in global_l2_errors
+      l2_error += error
+    end
+    linf_error = reduce((x, y) -> max.(x, y), global_linf_errors)
+
+    # For L2 error, divide by total volume
+    total_volume_ = total_volume(mesh)
+    l2_error = @. sqrt(l2_error / total_volume_)
+  else
+    l2_error = convert(T, NaN * zero(T))
+    linf_error = convert(T, NaN * zero(T))
+  end
 
   return l2_error, linf_error
 end
 
+function calc_error_norms_per_element(func, u, t, analyzer,
```

Comment:
I don't really like that we need to duplicate logic here. Why?

Reply:
I'm not sure what you mean here. Are you wondering why we can't reuse the functions for the serial case (via `invoke`, as in the original code) or are you wondering why I didn't put the contents of `calc_error_norms_per_element` directly into `calc_error_norms`?

Reply:
Basically resolved by "What's the motivation of this PR?"